source/elements/oneDNN/source/introduction.rst

```diff
@@ -0,0 +1,131 @@
+
+*Streams* (:struct:`dnnl::stream`) encapsulate execution context tied to a
+particular engine. For example, they can correspond to DPC++ command queues.
+
+Memory objects (:struct:`dnnl::memory`) encapsulate handles to memory
+allocated on a specific engine, tensor dimensions, data type, and memory
+format – the way tensor indices map to offsets in linear memory space. Memory
+objects are passed to primitives during execution.
+
+.. rubric:: Levels of Abstraction
+
+oneDNN has multiple levels of abstractions for primitives and memory objects
+in order to expose maximum flexibility to its users.
+
+On the logical level, the library provides the following abstractions:
+
+..
+   XXX
+   This is duplicated in primitives/index.rst and data_model/memory/index.rst
+
+* Memory descriptors (:struct:`dnnl::memory::desc`) define a tensor's logical
+  dimensions, data type, and the format in which the data is laid out in
+  memory. The special format any (:any:`dnnl::memory::format_tag::any`)
+  indicates that the actual format will be defined later.
+
+* Operation descriptors (one for each supported primitive) describe an
+  operation's most basic properties without specifying, for example, which
+  engine will be used to compute them. For example, convolution descriptor
+  describes shapes of source, destination, and weights tensors, propagation
+  kind (forward, backward with respect to data or weights), and other
+  implementation-independent parameters.
+
+* Primitive descriptors (:struct:`dnnl::primitive_desc_base` is the base class
+  and each of the supported primitives have their own version) are at an
+  abstraction level in between operation descriptors and primitives and can be
+  used to inspect details of a specific primitive implementation like expected
+  memory formats via queries to implement memory format propagation (see
+  Memory format propagation) without having to fully instantiate a primitive.
+
++--------------------------+-------------------+----------------------+
+| Abstraction level        | Memory object     | Primitive objects    |
++==========================+===================+======================+
+| Logical description      | Memory descriptor | Operation descriptor |
++--------------------------+-------------------+----------------------+
+| Intermediate description | N/A               | Primitive descriptor |
++--------------------------+-------------------+----------------------+
+| Implementation           | Memory object     | Primitive            |
++--------------------------+-------------------+----------------------+
+
+.. rubric:: General API notes
+
+There are certain assumptions on how oneDNN objects behave:
+
+- Memory and operation descriptors behave similarly to trivial types. Their
+  default constructors (only available for memory descriptors) create *zero*
+  objects that indicate absence of a memory object.  Copying memory and
```

Comment:
IMO, weird phrase: "their (mem and op) default ctors ... that indicate absence of a **memory** object", i.e. we started with op and mem, but finished with mems only.

Not even sure what you wanted to highlight here: the expensiveness of copying the objects, or the availability and behavior of _default_ constructors? Is it important?

Reply:
Reworded (deleted :))